### PR TITLE
Client directory

### PR DESCRIPTION
**Pull request by ReubenBond**
This PR replaces directory usage by clients (hosted and external) so that we can remove multiple activation support in the directory (see #6880).

The result is a client-specific directory which gossips client information around the cluster in a ring-fashion whereby each silo gossips to the next silo in the ring. Silos also directly query random silos when trying to lookup unknown clients.

The runtime interacts with this directory via the `ClientGrainLocator` implementation of `IGrainLocator`.

`ClientGrainLocator` relies on `ClientDirectory`, a new `SystemTarget` introduced by this PR which replaces `ClientObserverRegistrar`.

`ClientDirectory` maintains a mapping of clients connected to silos in the cluster. It sources information from:

* The local gateway, for which clients are locally connected and when they were last seen
* The cluster membership service, for which hosted clients exist in the cluster
* Other silos, via periodic updates and on-demand queries, to determine which clients are connected to other silos

### CODE REVIEW

File: src/Orleans.Core/Runtime/Constants.cs [hunk 1]
```diff
@@ -31,6 +31,7 @@ internal class Constants
         public static readonly GrainType DirectoryCacheValidatorType = SystemTargetGrainId.CreateGrainType("dir.cache-validator");
         public static readonly GrainType SiloControlType = SystemTargetGrainId.CreateGrainType("silo-control");
         public static readonly GrainType ClientObserverRegistrarType = SystemTargetGrainId.CreateGrainType("observer.registrar");
+        public static readonly GrainType ClientDirectoryType = SystemTargetGrainId.CreateGrainType("client-dir");
```
Comment (veikkoeeva) [inline src/Orleans.Core/Runtime/Constants.cs:34]: Out of curiosity, why `.`, dot, in some constants and `-`, a hyphen, in some?

Comment (ReubenBond) [reply src/Orleans.Core/Runtime/Constants.cs:34]: The original idea was to separate the category ("dir" "txn") from the component/implementation ("agent" "mem" "cache-validator"), but they appear inconsistent.

File: src/Orleans.Runtime/GrainDirectory/ILocalClientDirectory.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace Orleans.Runtime.GrainDirectory
+{
+    internal interface ILocalClientDirectory
+    {
+        bool TryLocalLookup(GrainId grainId, out List<ActivationAddress> addresses);
```
Comment (benjaminpetit) [inline src/Orleans.Runtime/GrainDirectory/ILocalClientDirectory.cs:8]: Since we have a `ClientGrainId` struct, I think we should use it in this interface

Comment (ReubenBond) [reply src/Orleans.Runtime/GrainDirectory/ILocalClientDirectory.cs:8]: The reason I didn't use that is that it's not a "transfer type" - i.e, I don't want it to exist over the wire, only within a single process

**Pull request merged**